src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt

```diff
@@ -308,35 +308,78 @@ internal open class ParallelThreadsRunner(
-                    afterInitStateRepresentation = afterInitStateRepresentation,
-                    afterParallelStateRepresentation = afterParallelStateRepresentation,
-                    afterPostStateRepresentation = afterPostStateRepresentation
-                )
-            )
+            return CompletedInvocationResult(collectExecutionResults(afterInitStateRepresentation, afterParallelStateRepresentation, afterPostStateRepresentation))
         } catch (e: TimeoutException) {
             val threadDump = collectThreadDump(this)
-            return RunnerTimeoutInvocationResult(threadDump)
+            return RunnerTimeoutInvocationResult(threadDump, collectExecutionResults())
         } catch (e: ExecutionException) {
-            return UnexpectedExceptionInvocationResult(e.cause!!)
+            return UnexpectedExceptionInvocationResult(e.cause!!, collectExecutionResults())
         } finally {
             resetState()
         }
     }
 
+    /**
+     * This method is called when we have some execution result other than [CompletedInvocationResult].
+     */
+    fun collectExecutionResults(): ExecutionResult {
+        return collectExecutionResults(null, null, null)
+    }
+
+    private fun collectExecutionResults(
+        afterInitStateRepresentation: String?,
+        afterParallelStateRepresentation: String?,
+        afterPostStateRepresentation: String?
+    ): ExecutionResult {
+        val initResults = initialPartExecution?.results?.let { collectExecutionResults(it, INIT) }.orEmpty()
+        val parallelResultsWithClock = parallelPartExecutions.map { execution ->
+            collectExecutionResults(execution.results, PARALLEL).zip(execution.clocks).map {
+                ResultWithClock(it.first, HBClock(it.second.clone()))
+            }
+        }
+        val postResults = postPartExecution?.results?.let { collectExecutionResults(it, POST) }.orEmpty()
+
+        return ExecutionResult(
+            initResults = initResults,
+            parallelResultsWithClock = parallelResultsWithClock,
+            postResults = postResults,
+            afterInitStateRepresentation = afterInitStateRepresentation,
+            afterParallelStateRepresentation = afterParallelStateRepresentation,
+            afterPostStateRepresentation = afterPostStateRepresentation
+        )
+    }
+
+    /**
+     * Composes a list of execution results from a raw array representation.
+     * When execution is finished in a normal way we have all non-nullable [TestThreadExecution.results].
+     * But when the execution has hung, results for some actors may be not set.
+     * This method checks if we have some unset results and if so, it replaces the first one as [NotFinished],
+     * and the rest as [NotStarted].
+     *
+     * @param executionPart indicates what execution part this result correspond to
+     */
+    private fun collectExecutionResults(rawResults: Array<Result?>, executionPart: ExecutionPart): List<Result> {
```

Comment:
I am still a bit confused. 

Why it is not simpler to just leave the `null`-s in the results?
The results array is initialized with `null`-s, and the result is set to non-null when the actor finishes. 
So `null` in the array indicates that the actor was not finished.